src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesVsPatterns.cs

```diff
@@ -0,0 +1,1336 @@
+            comp.VerifyDiagnostics(
+                // (12,18): warning CS8652: A default expression introduces a null value when 'T' is a non-nullable reference type.
+                //             t1 = default; // 1
+                Diagnostic(ErrorCode.WRN_DefaultExpressionMayIntroduceNullT, "default").WithArguments("T").WithLocation(12, 18),
+                // (15,9): warning CS8602: Possible dereference of a null reference.
+                //         t1.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t1").WithLocation(15, 9),
+                // (25,9): warning CS8602: Possible dereference of a null reference.
+                //         t2.ToString(); // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t2").WithLocation(25, 9)
+                );
+        }
+
+        [WorkItem(32503, "https://github.com/dotnet/roslyn/issues/32503")]
+        [Fact]
+        public void PatternDeclarationBreaksNullableAnalysis()
+        {
+            var source = @"
+#nullable enable
+class A { }
+class B : A
+{
+    A M()
+    {
+        var s = new A();
+        if (s is B b) {}
+        return s; 
+    }
+} 
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void InferenceWithITuplePattern()
+        {
+            var source = @"
+#nullable enable
+class A { }
+class B : A
+{
+    A M()
+    {
+        var s = new A();
+        if (s is B b) {}
+        return s; 
+    }
+} 
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void RecursivePatternNullInferenceWithDowncast_01()
+        {
+            var source = @"
+#nullable enable
+class Base
+{
+    public object Value = """";
+}
+class Derived : Base
+{
+    public new object Value = """";
+}
+class Program
+{
+    void M(Base? b)
+    {
+        if (b is Derived { Value: null })
+            b.Value.ToString();
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void RecursivePatternNullInferenceWithDowncast_02()
+        {
+            var source = @"
+#nullable enable
+class Base
+{
+    public object Value = """";
+}
+class Derived : Base
+{
+}
+class Program
+{
+    void M(Base? b)
+    {
+        if (b is Derived { Value: null })
+            b.Value.ToString(); // 1
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (15,13): warning CS8602: Possible dereference of a null reference.
+                //             b.Value.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b.Value").WithLocation(15, 13));
+        }
+
+        [Fact]
+        public void TuplePatternNullInference_01()
```

Comment:
Please expand the coverage of tuple patterns. I don't see any tests involving learning from non-null patterns, oblivious inputs being switched on, etc.